python/mxnet/ndarray/ndarray.py

```diff
@@ -507,71 +538,260 @@ def __getitem__(self, key):
+                        for dim_size in indices[i].shape[::-1]:
+                            k -= 1
+                            idx_rshape[k] = dim_size
                     else:
-                        raise ValueError(
-                            "NDArray does not support slicing with key %s of type %s."%(
-                                str(slice_i), str(type(slice_i))))
-                oshape.extend(shape[i+1:])
-                if len(oshape) == 0:
-                    oshape.append(1)
-                return op.slice(self, begin, end).reshape(oshape)
-        else:
-            raise ValueError(
-                "NDArray does not support slicing with key %s of type %s."%(
-                    str(key), str(type(key))))
+                        idx_rshape[basic_index_offset] = indices[i].shape[0]
+                        basic_index_offset += 1
+                    # broadcast current index to the final shape
+                    broadcasted_indices.append(indices[i].reshape(tuple(idx_rshape)).broadcast_to(index_bshape))
+                    # reset idx_rshape to ones
+                    for j, _ in enumerate(idx_rshape):
+                        idx_rshape[j] = 1
+
+            indices = broadcasted_indices
+
+        indices = op.stack(*indices)
+        return op.gather_nd(self, indices)
+
+    def _basic_indexing_step_equal_one(self, key):
+        """This function is called when key is a tuple of slices or integers with
+        all slices' steps being None or equal to 1."""
+        if not isinstance(key, tuple):
+            raise ValueError('index=%s must be a tuple type, received type=%s'
+                             % (str(key), str(type(key))))
+        shape = self.shape
+        oshape = []
+        begin = []
+        end = []
+        i = -1
+        for i, slice_i in enumerate(key):
+            if isinstance(slice_i, integer_types):
+                begin.append(slice_i)
+                end.append(slice_i+1)
+            elif isinstance(slice_i, py_slice):
+                # slice_i.step != 1 should not happen here
+                # it should have been routed to advanced indexing
+                if slice_i.step is not None and slice_i.step != 1:
+                    raise ValueError('_basic_indexing_step_equal_one does not support slice step=%d'
+                                     % slice_i.step)
+                begin.append(0 if slice_i.start is None else slice_i.start)
+                end.append(shape[i] if slice_i.stop is None else slice_i.stop)
+                oshape.append(end[i] - begin[i])
+            else:
+                raise ValueError('_basic_indexing_step_equal_one does not support slicing with '
+                                 'index=%s of type=%s.' % (str(slice_i), str(type(slice_i))))
+        oshape.extend(shape[i+1:])
+        if len(oshape) == 0:
+            oshape.append(1)
+        return op.slice(self, begin, end).reshape(oshape)
+
+    def _basic_indexing_step_not_one(self, key):
```

Comment:
This can be combined with _basic_indexing_step_equal_one by adding step to op.slice

Reply:
I will submit a separate PR implementing slice op with step to cover this change.